cmd/controller-manager/app/controller_manager.go

```diff
@@ -293,21 +293,37 @@ func StartControllers(s *options.ControllerManagerServer,
 	recorder record.EventRecorder,
 	stop <-chan struct{}) error {
 
-	// Get available service-catalog resources
-	glog.V(5).Info("Getting available resources")
-	availableResources, err := getAvailableResources(serviceCatalogClientBuilder)
-	if err != nil {
-		return err
-	}
-
-	coreKubeconfig = rest.AddUserAgent(coreKubeconfig, controllerManagerAgentName)
-	coreClient, err := kubernetes.NewForConfig(coreKubeconfig)
-	if err != nil {
-		glog.Fatal(err)
+	// When Catalog Controller and Catalog API Server are started at the
+	// same time with API Aggregation enabled, it may take some time before
+	// Catalog registration shows up in API Server.  Re-query several times if
+	// necessary.
+	var availableResources map[schema.GroupVersionResource]bool
+	catalogResourceAvailable := false
+	const maxAttempts = 4
```

Comment:
Is there a reason to bail out of the controller-manager at all instead of sitting in this loop with an exponential backoff indefinitely? Is there something that we gain from register a crash in the pod's status after 40 seconds that is better than registering it immediately as is done currently in master?

Reply:
I was hesitant to wait forever. there is nothing magic about 40 other then in my testing it may take upwards of 30 seconds to succeed when my machine is cpu challenged.